### PR TITLE
Parser for list of parameters

### PR DESCRIPTION
**Pull request by djbe**
Implements these issues:
- https://github.com/AliSoftware/SwiftGen/issues/253
- https://github.com/krzysztofzablocki/Sourcery/issues/14

This parses a list of strings (`Array<String>`) into a dictionary ([String: Any]). You can use it as follows:
```
let template = ...load template...
var context = ...generate context...

context["args"] = Parameters.parse(items: items)
let result = try template.render(context)
```

### CODE REVIEW

**Comment by djbe** [comment]
Is this what you guys had in mind? Should I add some sort of method for easily adding this to an (existing) context?

**Comment by djbe** [comment]
invalid keys?

File: Sources/Parameters.swift [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+		
+		var result = StringDict()
+		for parameter in parameters {
+			result = try parse(item: parameter, result: result)
+		}
+		
+		return result
+	}
+	
+	private static func parse(item: Parameter, result: StringDict) throws -> StringDict {
```
Comment (AliSoftware) [inline Sources/Parameters.swift]: Maybe rename that private function to `splitAndRepeat` or similar? Making it have the same name as the public one was confusing about its intend on first read

**Comment by AliSoftware** [comment]
I meant whatever use case the `ParameterError.invalidKey(key: String, value: String)` is for.

**Comment by djbe** [comment]
Yeah just saw that. Honestly can't remember when that'd happen. Which means the code needs comments!

**Comment by AliSoftware** [comment]
> Honestly can't remember when that'd happen. Which means the code needs comments!

And tests for those cases you thought about when writing that but forgot since then… making those cases likely to be forgotten again in the future and leading to regressions 😜

File: Sources/Parameters.swift [hunk 2]
```diff
@@ -7,53 +7,67 @@
...
+    let parameters: [Parameter] = try items.map { item in
+      let parts = item.components(separatedBy: "=")
+      guard parts.count == 2 else { throw ParametersError.invalidSyntax(value: item) }
+      return (key: parts[0], value: parts[1])
+    }
+    
+    var result = StringDict()
+    for parameter in parameters {
+      result = try parse(item: parameter, result: result)
+    }
```
Comment (AliSoftware) [inline Sources/Parameters.swift]: `parameters.reduce(StringDict()) { try Parse(item: $1, résultats: $0) }`?

File: Sources/Parameters.swift [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    let parameters: [Parameter] = try items.map { item in
+      let parts = item.components(separatedBy: "=")
+      guard parts.count == 2 else { throw ParametersError.invalidSyntax(value: item) }
+      return (key: parts[0], value: parts[1])
+    }
+    
+    var result = StringDict()
+    for parameter in parameters {
+      result = try parse(parameter: parameter, result: result)
+    }
```
Comment (AliSoftware) [inline Sources/Parameters.swift]: `parameters.reduce(StringDict()) { try Parse(item: $1, résultats: $0) }`?

Comment (djbe) [inline Sources/Parameters.swift]: That french autocorrect 😆

File: Sources/Parameters.swift [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+      let current = result[key] as? StringDict ?? StringDict()
+      let sub = (key: parts.suffix(from: 1).joined(separator: "."), value: parameter.value)
+      result[key] = try parse(parameter: sub, result: current)
+    }
+    
+    return result
+  }
+  
+  // a valid key is not empty and only alphanumerical
+  private static func validate(key: String) -> Bool {
```
Comment (AliSoftware) [inline Sources/Parameters.swift]: Does this means `1` is a valid key too?

File: Sources/Parameters.swift (hunk 3 above)
Comment (AliSoftware) [inline Sources/Parameters.swift]: 😂

File: Tests/TestSuites/ParametersTests.swift [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+  }
+  
+  func testInvalidSyntax() {
+    do {
+      let items = ["foo:1"]
+      _ = try Parameters.parse(items: items)
+    } catch ParametersError.invalidSyntax {
+      // That's the expected exception we want to happen
+    } catch let error {
+      XCTFail("Unexpected error occured while parsing: \(error)")
```
Comment (AliSoftware) [inline Tests/TestSuites/ParametersTests.swift]: Isn't there some `XCTAssertThrows` or similar?

File: Sources/Parameters.swift (hunk 4 above)
Comment (djbe) [inline Sources/Parameters.swift]: Right now, yeah. Shouldn't matter that much, say I want to create a (fake) tuple.

File: Tests/TestSuites/ParametersTests.swift (hunk 5 above)
Comment (djbe) [inline Tests/TestSuites/ParametersTests.swift]: No idea, I'll look into it. The structure is just based on the SwiftGenKit (colors) tests

File: Sources/Parameters.swift (hunk 4 above)
Comment (AliSoftware) [inline Sources/Parameters.swift]: Ok. Fine by be

File: Tests/TestSuites/ParametersTests.swift (hunk 5 above)
Comment (AliSoftware) [inline Tests/TestSuites/ParametersTests.swift]: That doesn't mean I was right at that time 😂

Comment (djbe) [inline Tests/TestSuites/ParametersTests.swift]: And I'm the testing noob 🤣

File: Sources/Parameters.swift [hunk 6]
```diff
@@ -19,8 +19,8 @@ public enum Parameters {
   public static func parse(items: [String]) throws -> StringDict {
     let parameters: [Parameter] = try items.map { item in
       let parts = item.components(separatedBy: "=")
-      if parts.count == 2 {
-        return (key: parts[0], value: parts[1])
+      if parts.count >= 2 {
+        return (key: parts[0], value: parts.dropFirst().joined(separator: "="))
       } else if let part = parts.first, parts.count == 1 && validate(key: part) {
```
Comment (AliSoftware) [inline Sources/Parameters.swift]: Does that mean that `foo.bar.baz` isn't a valid boolean key? (as `validate(key: "foo.bar.baz")` will return false)?

File: Sources/Parameters.swift [hunk 7]
```diff
@@ -66,9 +66,15 @@ public enum Parameters {
...
   
-  // a valid key is not empty and only alphanumerical
+  // a valid key is not empty and only alphanumerical or dot
   private static func validate(key: String) -> Bool {
     return !key.isEmpty &&
-      key.rangeOfCharacter(from: CharacterSet.alphanumerics.inverted) == nil
+      key.rangeOfCharacter(from: notAlfanumericsAndDot) == nil
   }
+  
+  private static let notAlfanumericsAndDot: CharacterSet = {
```
Comment (AliSoftware) [inline Sources/Parameters.swift]: s/alfa/alpha 😄

File: Tests/StencilSwiftKitTests/ParametersTests.swift [hunk 8]
```diff
@@ -86,6 +90,7 @@ class ParametersTests: XCTestCase {
   }
   
   func testInvalidKey() {
+    // key may only be alfanumeric or '.'
```
Comment (AliSoftware) [inline Tests/StencilSwiftKitTests/ParametersTests.swift]: s/alfa/alpha/

**Pull request merged**